contracts/src/WarpProtocolRegistry.sol

```diff
@@ -0,0 +1,195 @@
+    function getAddressFromVersion(
+        uint256 version
+    ) external view returns (address) {
+        return _getAddressFromVersion(version);
+    }
+
+    /**
+     * @dev Gets the version of the given `protocolAddress`.
+     * If `protocolAddress` is not a valid protocol address, returns 0, which is an invalid version.
+     */
+    function getVersionFromAddress(
+        address protocolAddress
+    ) external view returns (uint256) {
+        return _addressToVersion[protocolAddress];
+    }
+
+    /**
+     * @dev Gets the latest protocol version.
+     */
+    function getLatestVersion() external view returns (uint256) {
+        return _latestVersion;
+    }
+
+    /**
+     * @dev Gets and verifies for a warp out-of-band message, and adds the new protocol version
+     * addres to the registry.
+     */
+    function _addProtocolVersion(uint32 messageIndex) internal virtual {
+        // Get and validate for a warp out-of-band message.
+        (WarpMessage memory message, bool valid) = WARP_MESSENGER
+            .getVerifiedWarpMessage(messageIndex);
+        if (!valid) {
+            revert InvalidWarpMessage();
+        }
+        if (message.sourceChainID != _chainID) {
+            revert InvalidSourceChainID();
+        }
+        if (message.originSenderAddress != VALIDATORS_SOURCE_ADDRESS) {
+            revert InvalidOriginSenderAddress();
+        }
+        if (message.destinationChainID != _chainID) {
+            revert InvalidDestinationChainID();
+        }
+        if (message.destinationAddress != address(this)) {
+            revert InvalidDestinationAddress();
+        }
+
+        ProtocolRegistryEntry memory entry = abi.decode(
+            message.payload,
+            (ProtocolRegistryEntry)
+        );
+
+        _addToRegistry(entry);
+    }
+
+    /**
+     * @dev Adds the new protocol version address to the registry.
+     *
+     * Emits a {AddProtocolVersion} event when successful.
+     * Note: `protocolAddress` doesn't have to be a contract address, this is primarily
+     * to support the case we want to register a new protocol address meant for a security patch
+     * before the contract is deployed, to prevent the vulnerabilitiy from being exposed before registry update.
+     * Requirements:
+     *
+     * - `version` must be the increment of the latest version.
+     */
+    function _addToRegistry(
+        ProtocolRegistryEntry memory entry
+    ) internal virtual {
+        // Check that the version is the increment of the latest version.
+        if (entry.version != _latestVersion + 1) {
+            revert InvalidProtocolVersion();
+        }
+
+        if (entry.protocolAddress == address(0)) {
+            revert InvalidProtocolAddress();
+        }
+
+        _latestVersion++;
+        _versionToAddress[entry.version] = entry.protocolAddress;
+        _addressToVersion[entry.protocolAddress] = entry.version;
+        emit AddProtocolVersion(entry.version, entry.protocolAddress);
```

Comment:
What happens in the scenario where a new subnet is spun up after we're on - say - version 3. Are we guaranteed delivery of all the out-of-band messages? If so, are we guaranteed they're in order?

Reply:
In the constructor there is the ability to initialize the registry with version. So imo there are three options:
1) All existing versions of Teleporter are deployed to the chains, then we initialize the TeleporterRegistry with the same version->address mapping as other subnets.
2) If the subnet doesn't care for the previous existing Teleporter versions, they could have a different version mapping as other subnets. So the Teleporter at address A could be version 3 for subnet A, but is version 1 for subnet B
3) Subnet deploys existing Telepoter versions but don't initialize registry with mappings, then they'd need to update the registry with warp out of band messages, and choose whether they want the same ordering.

The first option would be the option I think we should advocate for, meaning all subnets would then have all version mappings to be the same, but for certain permissioned subnets, they may decide on their own registry mapping/orderings.

Reply:
I don't think we should make a subnet deploy every single version of teleporter if they don't want to.

I **really** don't think we should let different contracts have different versions on different chains. 

We should have a canonical mapping from version <-> address. What do we get from having the versions always start from 1 and being sequential?

Reply:
IIUC it's totally acceptable for a `TeleporterRegistry` on a new subnet to have the lowest version be greater than 1 - say it's 3 as Geoff said. Then all dApps that use the registry and that inherit `TeleporterUpgradeable` would have 3 as the minimum version anyway. Perhaps I'm misunderstanding something.

In any case, I don't think a canonical mapping is necessary, since ultimately it's the Teleporter address that determines if the message can be delivered. With this `TeleporterUpgradeable` pattern, the versions are used more as a backwards counter from the latest Teleporter version, rather than a forwards counter from the initial Teleporter version. For example, say a dApp on subnetA and subnetB upgrade from Teleporter 0x123 to 0xabc, but still allow Teleporter 0x123 by not calling `updateMinTeleporterVersion`. All the registry logic should still work, even if the subnets assign different integer versions to the same Teleporter contract address.

Reply:
If we decide to go with the approach where all new subnets start from Teleporter version 1 internally, regardless of the global Teleporter version, we could clean up `WarpProtocolRegistry.sol` by not having to include the version number in `_addToRegistry`, as well as not having to include it in the out-of-band message.

Reply:
Right now the `TeleporterRegistry` initialization requires that the version starts with 1 AND is incremented, meaning all Teleployer versions would need to be deployed and specified in order. I'm thinking for the initialization to remove these requirements, so you can specify any mapping and doesn't have to be in incremental order. For example if there versions 1,2,3,4,5, and a subnet wanted to just deploy 2,4,5 they can. 

The next question is after the registry is initialized should we require that versions are in order. For a subnet to have the same mapping after initialization as other subnets, it would need to register every new Teleporter version to the registry. How about for subnets that don't want to deploy every version? They would end up with a mismatched registry mapping.

Another solution is to not require incremental versions altogether, so you can add whatever version->address mapping, that way a subnet that doesn't want to deploy a certain version can skip it and not register in the registry. We'd just need to handle cases like if we're adding a version < _latestVersion that we don't update _latestVersion, etc.

Reply:
A completely different version mapping would be a set canonical mapping of version->protocolAddress. This could be done by adding a `ITeleporterMessenger.getVersion()`, and a hardcoded constant `VERSION` in `TeleporterMessenger`. So when we want to update registry we pass in the new protocol address, cast and get version `ITeleporterMessenger(protocolAddress).getVersion()`. 

There is a case lets say we have versions 1,2,3, but version 3 has a bug, and we want dapps to use version 2. Then the previous `_minTeleporterVersion` of dapps will stay at 3 while `_latestVersion` is 2. I think there are two ways around this:
1) In `onlyAllowedTeleporter` if  `_minTeleporterVersion > _latestVersion` then we only allow addres of `_latestVersion`.
2) We tell registry to add address for version 2, registry sees that version 2 already is registered, then registers this address as a new version so we have 1,2,3,4 now. The next contract we make should skip 4 and be version 5. We may want separate registry add function for these rollback cases.

Reply:
Overall leaning towards not the canonical mapping of registry between subnets. We can remove the requirement to start from version 1 in initialization. We'd be leaving the mapping order of version->protocolAddress up to subnets and their out-of-band warp messages, but still recommend to maintain the same ordering that we push out for versions. Curious to hear thoughts.

Reply:
I have some concerns about mismatched version numbers between different subnets. I understand that it ultimately doesn't matter since only the address matters, but it can lead to some confusion/misunderstanding that can result in incorrect actions in the event of an emergency upgrade. 

What are the reasons to not have the hardcoded VERSION constant that you had suggested?

Reply:
My main concern with a canonical version across subnets was how to guarantee the mapping of version to address was consistent across all subnets. Hardcoding the version in `TeleporterMessenger` would achieve that, as well as reducing the change for confusion as @bernard-avalabs mentioned.

That's an interesting idea for the rollback case. I'm trying to think if there are any odd edge cases that that could introduce. Say we want to rollback to version `2` at address `0x123`, and we're currently on version `3`. The internal mapping would then be `2->0x123` and `4->0x123`. In the future, it we need to rollback to `0x123` again, then either version `2` or `4` would be able to trigger that. Does this present any issues if, say, one subnet rolls back to `0x123` using version `2` while another rolls back using version `4`? Similarly, are there any compatability issues if different subnets have the same address registered at different versions due to inconsistent rollbacks? I think the answer to both is no, but we should consider similar cases.

Reply:
Discussed offline, we are going with an approach that removes the requirements for initialization to start from version 1, and for versions added to the registry to be incremental. When versions that are greater than latest version are added, we update the latest version. But users can also add old versions, as long as they were not previously registered, and the registry will simply add the entry without updating latest version.